Review N+1 (#2)
biohazard999/Scissors.Blog: _drafts/2020-09-17-fixing-an-n1-performance-problem-in-xaf-xpo-with-totally-undocumented-apis.md

---
 layout: post 
 title: Fixing an N+1 performance problem in XAF/XPO with totally undocumented APIs
 comments: true
 github: fixing-an-n-plus-1-perf-problem-in-xaf-xpo
 tags: ["XAF", "XPO", "Performance", "BestPractices", "ORM"]
---

We all know (or not) that using an abstraction over any kind of datasource can lead to performance problems. That's nothing new and is quite common with ORM's.

Let's look at the business problem we need to solve first before discussing why N+1 happens and what we can do to avoid it.
It's a quite common problem with relational databases, an especially when dealing with any ORM. So we need to be careful to avoid it and there are several options we have when dealing with that kind of problem.

## The scenario (is quite artificial but will serve the purpose)

> Okay to be clear, this is an rather easy one and you don't want to go all the way down this performance optimization route (esp. for this example) but it's universial, easy to understand and applies to all 1+N aggregate calculation problems.

Reviewer: ```suggestion
Okay to be clear, this is an rather easy one and you don't want to go all the way down this performance optimization route (esp. for this example) but it's universal, easy to understand and applies to all 1+N aggregate calculation problems.
```

Let's imagine you have a simple `Offer` and `OfferItem` class that are in a `1 + N` relationship.

```txt
  Offer                       OfferItem
+-----------------+         +-----------------+
| OfferId         |         | OfferItemId     |
| Name (str)      |         |                 |
|                 |         |  Hours (int)    |
|                 +-------->+  FK_Offer       |
|                 |         |                 |
|                 |         |                 |
+-----------------+         +-----------------+
```

> So let's imagine we want to calculate the sum of the hour's and display them in the offer's `ListView`. We have now several options and I will step through them from **naive** (slow) till **crazy** (fast). The last couple of techniques use some internals of XPO and are *not supported* by XPO, but I found them so useful so it would be a shame not not mention them.

Reviewer: hour's ->  hours

> So let's imagine we want to calculate the sum of the hour's and display them in the offer's `ListView`. We have now several options and I will step through them from **naive** (slow) till **crazy** (fast). The last couple of techniques use some internals of XPO and are *not supported* by XPO, but I found them so useful so it would be a shame not not mention them.

Reviewer: till -> to

> So let's imagine we want to calculate the sum of the hour's and display them in the offer's `ListView`. We have now several options and I will step through them from **naive** (slow) till **crazy** (fast). The last couple of techniques use some internals of XPO and are *not supported* by XPO, but I found them so useful so it would be a shame not not mention them.

Reviewer: found -> find

> > Be warned! Some of them will probably break in the future, or don't work under all circumstances. I will mark them with an disclaimer.

Reviewer: an disclaimer -> a disclaimer

> > Be warned! Some of them will probably break in the future, or don't work under all circumstances. I will mark them with an disclaimer.

Reviewer: ```suggestion
> Be warned! Some of them will probably break in the future, or don't work under all circumstances. I will mark them with a disclaimer.
```

But first let's talk about what's the N+1 problem is all about!

### The N+1 problem

> Object Relational Mappers (or ORM's for short) are a pattern to abstract away the database access and *project* SQL queries to objects. In our case that will be XPO that translate our access to the database.

Reviewer: that will be XPO -> it will be XPO

> Object Relational Mappers (or ORM's for short) are a pattern to abstract away the database access and *project* SQL queries to objects. In our case that will be XPO that translate our access to the database.

Reviewer: ```suggestion
Object Relational Mappers (or ORM for short) are a pattern to abstract away the database access and *project* SQL queries to objects. In our case it will be XPO that translated our access to the database.
```

> To tell XPO about our database we need to tell it some information about the database and it's relationship between entities with attributes. That allows XPO to *guess* what SQL statements it should generate. This is a very powerful abstraction to have, cause you don't have to think all the time about SQL and can focus on business logic. That's fine for the most part but if the number of records grow (or tables and relationships get more complicated) that guess can horrible go wrong, or even worse, you give the wrong hint's to the ORM so it only can perform multiple queries to the database.

Reviewer: ```suggestion
To tell XPO about our database, we use attributes to specify relationships between entities. That allows XPO to *guess* what SQL statements it should generate. This is a very powerful abstraction, because you don't need to constantly think about SQL and can focus on business logic. That's fine for the most part but if the number of records grow (or tables and relationships get more complicated) that guess can go horribly wrong.
```
I removed the last sentence, because what can be worse than something that goes horribly wrong?

> Expensive queries are also something that can occur with ORM's (like massive JOIN's) but that's not the focus of this blog post.

Reviewer: ORM's -> ORMs

> Expensive queries are also something that can occur with ORM's (like massive JOIN's) but that's not the focus of this blog post.

Reviewer: ```suggestion
Expensive queries are also something that can occur with ORMs (like massive JOIN's) but that's not the focus of this blog post.
```

---
 layout: post 
 title: Fixing an N+1 performance problem in XAF/XPO with totally undocumented APIs
 comments: true
 github: fixing-an-n-plus-1-perf-problem-in-xaf-xpo
 tags: ["XAF", "XPO", "Performance", "BestPractices", "ORM"]
---

> We all know (or not) that using an abstraction over any kind of datasource can lead to performance problems. That's nothing new and is quite common with ORM's.

Reviewer: ```suggestion
We all know (or not) that using an abstraction over any kind of datasource can lead to performance problems. That's nothing new and is quite common with ORMs.
```

Let's look at the business problem we need to solve first before discussing why N+1 happens and what we can do to avoid it.
It's a quite common problem with relational databases, an especially when dealing with any ORM. So we need to be careful to avoid it and there are several options we have when dealing with that kind of problem.

## The scenario (is quite artificial but will serve the purpose)

Okay to be clear, this is an rather easy one and you don't want to go all the way down this performance optimization route (esp. for this example) but it's universal, easy to understand and applies to all 1+N aggregate calculation problems.

Let's imagine you have a simple `Offer` and `OfferItem` class that are in a `1 + N` relationship.

```txt
  Offer                       OfferItem
+-----------------+         +-----------------+
| OfferId         |         | OfferItemId     |
| Name (str)      |         |                 |
|                 |         |  Hours (int)    |
|                 +-------->+  FK_Offer       |
|                 |         |                 |
|                 |         |                 |
+-----------------+         +-----------------+
```

> So let's imagine we want to calculate the sum of the hour's and display them in the offer's `ListView`. We have now several options and I will step through them from **naive** (slow) till **crazy** (fast). The last couple of techniques use some internals of XPO and are *not supported* by XPO, but I found them so useful so it would be a shame not not mention them.

Reviewer: ```suggestion
We want to calculate the sum of the hours and display them in the offer's `ListView`. We have now several options and I will step through them from **naive** (slow) to **crazy** (fast). The last couple of techniques use some internals of XPO and are *not supported* by XPO, but I found them so useful so it would be a shame not not mention them.
```

> > Be warned! Some of them will probably break in the future, or don't work under all circumstances. I will mark them with a disclaimer.

Reviewer: ```suggestion
> Be warned! Some of them will probably break in the future, or won't work under all circumstances. I will mark them with a disclaimer.
```

But first let's talk about what's the N+1 problem is all about!

### The N+1 problem

Object Relational Mappers (or ORM for short) are a pattern to abstract away the database access and *project* SQL queries to objects. In our case it will be XPO that translates our access to the database.

> To tell XPO about our database we need to tell it some information about the database and it's relationship between entities with attributes. That allows XPO to *guess* what SQL statements it should generate. This is a very powerful abstraction to have, because you don't have to think all the time about SQL and can focus on business logic. That's fine for the most part but if the number of records grow (or tables and relationships get more complicated) that guess can horrible go wrong, or even worse, you give the wrong hint's to the ORM so it only can perform multiple queries to the database.

Reviewer: ```suggestion
We tell XPO about our database and its relationship between entities using attributes. That allows XPO to *guess* what SQL statements it should generate. This is a very powerful abstraction, because you don't have to think all the time about SQL and can focus on business logic. That's fine for the most part, but if the number of records grows (or tables and relationships get more complicated) that guess can go horribly wrong, or even worse, you give the wrong hints to the ORM and it it performs multiple superfluous queries to the database.
```

> To tell XPO about our database, we use attributes to specify relationships between entities. That allows XPO to *guess* what SQL statements it should generate. This is a very powerful abstraction, because you don't need to constantly think about SQL and can focus on business logic. That's fine for the most part but if the number of records grow (or tables and relationships get more complicated) that guess can go horribly wrong.

Reviewer: Repeat of previous paragraph? I guess you can choose whichever you prefer. Personally I prefer the second one.

Author: Damn merge conflicts. My eyes burn. Okay lets go for the second one

> Expensive queries are also something that can occur with ORMs (like massive JOIN's) but that's not the focus of this blog post.

Reviewer: ```suggestion
Expensive queries are also something that can occur with ORMs (like massive JOINs) but that's not the focus of this blog post.
```

I'll give you a litte example in **very** naive C#.

> > PLEASE NEVER DO SOMETHING LIKE THIS IN PRODUCTION. YOU HAVE BEEN WARNED. (Or I insist you NEED to [book some consulting hour's](https://www.delegate.at) from me)

Reviewer: ```suggestion
> PLEASE NEVER DO SOMETHING LIKE THIS IN PRODUCTION. YOU HAVE BEEN WARNED. (Or I insist you NEED to [book some consulting hours](https://www.delegate.at) with me)
```

```cs
[DefaultClassOptions]
[DefaultProperty(nameof(HourSum))]
public class SlowOffer : BaseObject
{
    public SlowOffer(Session session) : base(session) { }

    private string _Name;
    [Persistent("Name")]
    public string Name { get => _Name; set => SetPropertyValue(nameof(Name), ref _Name, value); }

    [NonPersistent]
    public int HourSum
    {
        get
        {
            var sum = 0;
            foreach (var item in OfferItems)
            {
                sum += item.Hours;
            }
            return sum;
        }
    }

    [Association, Aggregated]
    public XPCollection<SlowOfferItem> OfferItems => GetCollection<SlowOfferItem>(nameof(OfferItems));
}

public class SlowOfferItem : BaseObject
{
    public SlowOfferItem(Session session) : base(session) { }

    private int _Hours;
    [Persistent("Hours")]
    public int Hours { get => _Hours; set => SetPropertyValue(nameof(Hours), ref _Hours, value); }

    private SlowOffer _SlowOffer;
    [Persistent("SlowOffer"), Association]
    public SlowOffer SlowOffer { get => _SlowOffer; set => SetPropertyValue(nameof(SlowOffer), ref _SlowOffer, value); }
}

```

> > For all examples I use 300 `Offers` with 1000 `OfferItems` each. It's sorted by the `HourSum` property acending. I use SqlServer LocalDb for this tests on a Intel(R) Core(TM) i7-4770K CPU @ 3.50GHz with 16 GB of memory and a two SSD disks Raid 0 setup.  

Reviewer: ```suggestion
> For all examples I use 300 `Offers` with 1000 `OfferItems` each. It's sorted by the `HourSum` property ascending. I use SqlServer LocalDb for these tests on a Intel(R) Core(TM) i7-4770K CPU @ 3.50GHz with 16 GB of memory and a two SSD disks Raid 0 setup.  
```

The build configuration is set on Debug with an attached debugger.  
XPO log verbosity = 4  
XAF log verbosity = 3  

> This means those numbers *will increase in Release* configuration, so we setting the base line pessimistic (and also aim for an better developer feedback cycle).

Reviewer: ```suggestion
This means those numbers *will improve in Release* configuration, so we set a pessimistic base line (and also aim for a better developer feedback cycle).
```

I'm guessing you mean that the timed values will decrease in RELEASE, so the 'numbers' will improve.

18.09.20 16:22:20.053 Result: rowcount = 1000, total = 44000, N0.{Oid,Guid} = 16000, N0.{Hours,Int32} = 4000, N0.{SlowOffer,Guid} = 16000, N0.{OptimisticLockField,Int32} = 4000, N0.{GCRecord,Int32} = 4000
18.09.20 16:22:20.057 Executing sql 'select N0."Oid",N0."Hours",N0."SlowOffer",N0."OptimisticLockField",N0."GCRecord" from "dbo"."SlowOfferItem" N0 where (N0."GCRecord" is null and (N0."SlowOffer" = @p0))' with parameters {3b48cf0a-d610-4586-8dfb-fde802777653}
18.09.20 16:22:20.078 Result: rowcount = 1000, total = 44000, N0.{Oid,Guid} = 16000, N0.{Hours,Int32} = 4000, N0.{SlowOffer,Guid} = 16000, N0.{OptimisticLockField,Int32} = 4000, N0.{GCRecord,Int32} = 4000
18.09.20 16:22:20.122 Executing sql 'select top 1 count(*) from "dbo"."SlowOffer" N0 where N0."GCRecord" is null'
18.09.20 16:22:20.126 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
18.09.20 16:22:20.127 Executing sql 'select top 1 count(*) from "dbo"."SlowOfferItem" N0 where N0."GCRecord" is null'
18.09.20 16:22:20.170 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
```

> What happens is: for every row in the `Orders` table a second select that selects all the related `OrderItems` collection get selected one by one. So in total there are at least 300 single `SELECT * from OrderItem where OrderId = X` statements dropped at your database.

Reviewer: ```suggestion
What happens is: for every row in the `Offer` table a second select that selects all the related `OfferItems` collection get selected one by one. So in total there are at least 300 single `SELECT * from OfferItem where OfferId = X` statements dropped at your database.
```

19.09.20 10:35:36.803 Result: rowcount = 1000, total = 44000, N0.{Oid,Guid} = 16000, N0.{Hours,Int32} = 4000, N0.{SlowOfferWithLinq,Guid} = 16000, N0.{OptimisticLockField,Int32} = 4000, N0.{GCRecord,Int32} = 4000
19.09.20 10:35:36.851 Executing sql 'select top 1 count(*) from "dbo"."SlowOfferWithLinq" N0 where N0."GCRecord" is null'
19.09.20 10:35:36.853 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
19.09.20 10:35:36.854 Executing sql 'select top 1 count(*) from "dbo"."SlowOfferItemWithLinq" N0 where N0."GCRecord" is null'
19.09.20 10:35:36.877 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
```

This results in the following performance

1. Sort by `Name`: 0.735 seconds
2. Sort by `HourSum`: 5.672 seconds

As you can see, it seams like it performs slightly faster, but on average it is exactly the same as the naive approach.  
The reason for this is hard to spot (and that is why *lazy loading can be dangerous*, if you don't know what you are doing):

```cs
[NonPersistent]
// Sum up all the hours using Linq
public int HourSum => OfferItems.Sum(m => m.Hours);
```

> We first load the entire collection and sum up after wards in memory. The reason for this is pretty simple by looking at the datatype of `OfferItems`.

Reviewer: ```suggestion
We first load the entire collection and sum up afterwards in memory. The reason for this is pretty simple by looking at the datatype of `OfferItems`.
```

It's a `XPCollection<SlowOfferItemWithLinq>`. In order to let the database the work, we need to make sure we work with an `IQueryable<SlowOfferItemWithLinq>`.

### Linq to the rescue with IQueryable

> Know we know what we should aim for let's look at a Linq version of the code that uses an `IQueryable`:

Reviewer: ```suggestion
Now we know what we should aim for, let's look at a Linq version of the code that uses an `IQueryable`:
```

19.09.20 10:58:05.075 Result: rowcount = 1, total = 4, SubQuery(Sum,N0.{Hours,Int32},) = 4
19.09.20 10:58:05.076 Executing sql 'select top 1 sum(N0."Hours") from "dbo"."FasterOfferItemWithLinq" N0 where (N0."GCRecord" is null and (N0."FasterOfferWithLinq" = @p0))' with parameters {3683bfc2-7314-452f-be51-00692f123965}
.
.
.
19.09.20 10:58:06.497 Executing sql 'select top 1 count(*) from "dbo"."FasterOfferWithLinq" N0 where N0."GCRecord" is null'
19.09.20 10:58:06.497 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
19.09.20 10:58:06.498 Executing sql 'select top 1 count(*) from "dbo"."FasterOfferItemWithLinq" N0 where N0."GCRecord" is null'
19.09.20 10:58:06.512 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
19.09.20 10:58:07.919 Executing sql 'select top 1 sum(N0."Hours") from "dbo"."FasterOfferItemWithLinq" N0 where (N0."GCRecord" is null and (N0."FasterOfferWithLinq" = @p0))' with parameters {820161dc-8568-412c-bf84-c4f148e8122f}
19.09.20 10:58:07.924 Result: rowcount = 1, total = 4, SubQuery(Sum,N0.{Hours,Int32},) = 4
```

This results in the following performance

1. Sort by `Name`: 0.156 seconds
2. Sort by `HourSum`: 1.554 seconds

> As you can see, it performs a lot better. And for the *most parts* of your application this will be totally fine (esp. if you arn't dealing with too much records, or lists that arn't used that frequently).

Reviewer: ```suggestion
As you can see, it performs a lot better. And for the *most parts* of your application, this will be totally fine (esp. if you aren't dealing with too many records, or lists that aren't used that frequently).
```

.
.
.
19.09.20 11:20:33.695 Executing sql 'select N0."Oid",N0."Hours",N0."SlowOfferWithPersistentAlias",N0."OptimisticLockField",N0."GCRecord" from "dbo"."SlowOfferItemWithPersistentAlias" N0 where (N0."GCRecord" is null and (N0."SlowOfferWithPersistentAlias" = @p0))' with parameters {926fde40-af9c-4679-bdf9-fe0edcc49465}
19.09.20 11:20:33.718 Result: rowcount = 1000, total = 44000, N0.{Oid,Guid} = 16000, N0.{Hours,Int32} = 4000, N0.{SlowOfferWithPersistentAlias,Guid} = 16000, N0.{OptimisticLockField,Int32} = 4000, N0.{GCRecord,Int32} = 4000
19.09.20 11:20:33.789 Executing sql 'select top 1 count(*) from "dbo"."SlowOfferWithPersistentAlias" N0 where N0."GCRecord" is null'
19.09.20 11:20:33.790 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
19.09.20 11:20:33.790 Executing sql 'select top 1 count(*) from "dbo"."SlowOfferItemWithPersistentAlias" N0 where N0."GCRecord" is null'
19.09.20 11:20:33.805 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
```

This results in the following performance

1. Sort by `Name`: 0.671 seconds
2. Sort by `HourSum`: 6.501 seconds

> But why is that? XAF will load data according to the `DataAccessMode` property of the `ListView`. Until now we always used the `Client` mode. Which is fine for the most part. It's the most convinient to use cause you *almost never*  need to think about the underlying database.  

Reviewer: ```suggestion
But why is that? XAF will load data according to the `DataAccessMode` property of the `ListView`. Until now we always used the `Client` mode. Which is fine for the most part. It's the most convenient to use cause you *almost never*  need to think about the underlying database.  
```

That also means, it will do the calculation client side, which will result in the same queries as we know from the *naive* approach.

### PersistentAlias performs better with the correct DataAccessMode

> We have several options [provided by XAF](https://docs.devexpress.com/eXpressAppFramework/113683/concepts/ui-construction/views/list-view-data-access-modes) to change the behavior how data will be loaded. I will list them in the order they where introduced in the framework. Every mode has it's own strenghts and weaknesses. I'll only look into the performance right now. Programming model varies between those and i will skip `Client` cause we looked at it before. The source code is exactly the same before. To change the `DataAccessMode` we change it's property in the `ModelEditor`.

Reviewer: ```suggestion
We have several options [provided by XAF](https://docs.devexpress.com/eXpressAppFramework/113683/concepts/ui-construction/views/list-view-data-access-modes) to change the behavior how data will be loaded. I will list them in the order they were introduced in the framework. Every mode has its own strengths and weaknesses. I'll only look into the performance right now. Programming model varies between those and i will skip `Client` cause we looked at it before. The source code is exactly the same before. To change the `DataAccessMode` we change it's property in the `ModelEditor`.
```

```txt
19.09.20 12:03:31.738 Executing sql 'select top 2147483647 (select sum(N1."Hours") as Res0 from "dbo"."SlowOfferItemWithPersistentAlias" N1 where ((N0."Oid" = N1."SlowOfferWithPersistentAlias") and N1."GCRecord" is null)),N0."Name",N0."Oid" from "dbo"."SlowOfferWithPersistentAlias" N0 where N0."GCRecord" is null'
19.09.20 12:03:31.821 Result: rowcount = 300, total = 10848, SubQuery(Sum,N1.{Hours,Int32},Select N1.{Hours,Int32}
  from "SlowOfferItemWithPersistentAlias" N1
 where N0.{Oid,Guid} = N1.{SlowOfferWithPersistentAlias,Guid} And N1.{GCRecord,Int32} Is Null) = 1200, N0.{Name,String} = 4848, N0.{Oid,Guid} = 4800
19.09.20 12:03:31.824 Executing sql 'select top 1 count(*) from "dbo"."SlowOfferWithPersistentAlias" N0 where N0."GCRecord" is null'
19.09.20 12:03:31.825 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
19.09.20 12:03:31.825 Executing sql 'select top 1 count(*) from "dbo"."SlowOfferItemWithPersistentAlias" N0 where N0."GCRecord" is null'
19.09.20 12:03:31.839 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
```

This results in the following performance

1. Sort by `Name`: 0.071 seconds
2. Sort by `HourSum`: 0.078 seconds

#### InstantFeedback

> In `InstantFeedback` mode records are loaded in an async manor. It operates almost the same as the `ServerMode` in terms of number of queries, but the user experience is much better, cause visible records will be fetched in the background.

Reviewer: ```suggestion
In `InstantFeedback` mode records are loaded in an asynchronously . It operates almost the same as the `ServerMode` in terms of number of queries, but the user experience is much better, because visible records will be fetched in the background.
```

19.09.20 12:08:07.960 Result: rowcount = 1000, total = 44000, N0.{Oid,Guid} = 16000, N0.{Hours,Int32} = 4000, N0.{SlowOfferWithPersistentAlias,Guid} = 16000, N0.{OptimisticLockField,Int32} = 4000, N0.{GCRecord,Int32} = 4000
.
.
.
19.09.20 12:08:08.704 Executing sql 'select N0."Oid",N0."Hours",N0."SlowOfferWithPersistentAlias",N0."OptimisticLockField",N0."GCRecord" from "dbo"."SlowOfferItemWithPersistentAlias" N0 where (N0."GCRecord" is null and (N0."SlowOfferWithPersistentAlias" = @p0))' with parameters {1a57b4dc-7812-48cc-83ae-15e184078c5f}
19.09.20 12:08:08.718 Result: rowcount = 1000, total = 44000, N0.{Oid,Guid} = 16000, N0.{Hours,Int32} = 4000, N0.{SlowOfferWithPersistentAlias,Guid} = 16000, N0.{OptimisticLockField,Int32} = 4000, N0.{GCRecord,Int32} = 4000
```

This results in the following performance

First user interaction:

1. Sort by `Name`: 0.040 seconds
2. Sort by `HourSum`: 0.030 seconds

Visible data loaded:

1. Sort by `Name`: ~~0.800 seconds
2. Sort by `HourSum`: ~~0.800 seconds

### InstantFeedbackView

> In `InstantFeedbackView` mode records are loaded in an async manor. It operates almost the same as the `DataView` mode in terms of number of queries, but the user experience **can be** much better, cause all records will be fetched in the background.

Reviewer: ```suggestion
In `InstantFeedbackView` mode records are loaded in an asynchronously. It operates almost the same as the `DataView` mode in terms of number of queries, but the user experience **can be** much better, cause all records will be fetched in the background.
```

Author: Shouldnt that be: loaded asynchronosly? Minus *in an*

19.09.20 12:13:23.524 Result: rowcount = 55, total = 4236, N0.{Name,String} = 908, N0.{Oid,Guid} = 880, SubQuery(Sum,N1.{Hours,Int32},Select N1.{Hours,Int32}
  from "SlowOfferItemWithPersistentAlias" N1
 where N0.{Oid,Guid} = N1.{SlowOfferWithPersistentAlias,Guid} And N1.{GCRecord,Int32} Is Null) = 220, SubQuery(Sum,N2.{Hours,Int32},Select N2.{Hours,Int32}
  from "SlowOfferItemWithPersistentAlias" N2
 where N0.{Oid,Guid} = N2.{SlowOfferWithPersistentAlias,Guid} And N2.{GCRecord,Int32} Is Null) = 220, N0.{Name,String} = 908, N0.{Oid,Guid} = 880, SubQuery(Sum,N3.{Hours,Int32},Select N3.{Hours,Int32}
  from "SlowOfferItemWithPersistentAlias" N3
 where N0.{Oid,Guid} = N3.{SlowOfferWithPersistentAlias,Guid} And N3.{GCRecord,Int32} Is Null) = 220
```

This results in the following performance

First user interaction:

1. Sort by `Name`: 0.006 seconds
2. Sort by `HourSum`: 0.007 seconds

Visible data loaded:

1. Sort by `Name`: ~~0.0300 seconds
2. Sort by `HourSum`: ~~0.400 seconds

### ServerView

> In `ServerView` nide only a fixed amount of records are fetched from the database, as well as including the sorting query and some support queries. Afterwards visible columns are fetched on demand similar to the `DataView` but with pagination and visible columns support.

Reviewer: ```suggestion
In `ServerView` only a fixed amount of records are fetched from the database, as well as including the sorting query and some support queries. Afterwards visible columns are fetched on demand similar to the `DataView` but with pagination and visible columns support.
```

Author: Nide=Mode. ?

  from "SlowOfferItemWithPersistentAlias" N1
 where N0.{Oid,Guid} = N1.{SlowOfferWithPersistentAlias,Guid} And N1.{GCRecord,Int32} Is Null) = 220, N0.{Name,String} = 908, N0.{Oid,Guid} = 880, N0.{Oid,Guid} = 880, SubQuery(Sum,N2.{Hours,Int32},Select N2.{Hours,Int32}
  from "SlowOfferItemWithPersistentAlias" N2
 where N0.{Oid,Guid} = N2.{SlowOfferWithPersistentAlias,Guid} And N2.{GCRecord,Int32} Is Null) = 220
19.09.20 12:19:10.995 Executing sql 'select top 1 count(*) from "dbo"."SlowOfferWithPersistentAlias" N0 where N0."GCRecord" is null'
19.09.20 12:19:10.995 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
19.09.20 12:19:10.996 Executing sql 'select top 1 count(*) from "dbo"."SlowOfferItemWithPersistentAlias" N0 where N0."GCRecord" is null'
19.09.20 12:19:11.010 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
```

This results in the following performance

1. Sort by `Name`: 0.079 seconds
2. Sort by `HourSum`: 0.236 seconds

> **Important**: When using `PersistentAlias` for aggregate calculations always consider using a `DataAccessMode` that can perform database side calculation.  

> There is no silverbullet and you need to measure careful in different configurations what fit's best.

Reviewer: ```suggestion
There is no silver bullet and you need to measure carefully using different configurations to find what is the best fit.
```

All comes with a cost. All data access modes have [some limitations](https://docs.devexpress.com/eXpressAppFramework/113683/concepts/ui-construction/views/list-view-data-access-modes#non-persistent-object-support-limitations). If you need grouping and sorting capabilities by some fields (or even just display `NonPersistent` members) you are *out of luck*. It is also *may* be limited to the complexity you can fit into a single criteria.  

## Techniques that overcome the N+1 problems

> Now we learned about the N+1 problem and some useful ways to overcome some of the performance problems that XAF provides we will cover 2 supported and 1 unsupported techniques how to avoid the N+1 problem completely.  

Reviewer: ```suggestion
Now we learned about the N+1 problem and some useful ways to overcome some of the performance problems that XAF causes. We will cover 2 supported and 1 unsupported techniques to avoid the N+1 problem completely.  
```

Author: *can* cause.
XAF cant know what we want. Poor XAF ❤️

    public FasterOfferWithCQRS FasterOfferWithCQRS { get => _FasterOfferWithCQRS; set => SetPropertyValue(nameof(FasterOfferWithCQRS), ref _FasterOfferWithCQRS, value); }
}

[DefaultProperty(nameof(HourSum))]
public class FasterOfferWithCQRSQuery : BaseObject
{
    public FasterOfferWithCQRSQuery(Session session) : base(session) { }

    private FasterOfferWithCQRS _FasterOfferWithCQRS;
    [Persistent("FasterOfferWithCQRS")]
    [MemberDesignTimeVisibility(false)] //We don't want to see the data in the view, we only want the reference as a foreign key
    //Save the reference for comparison. We also can use a unique index on this if needed, but we omit that for simplicity
    public FasterOfferWithCQRS FasterOfferWithCQRS { get => _FasterOfferWithCQRS; set => SetPropertyValue(nameof(FasterOfferWithCQRS), ref _FasterOfferWithCQRS, value); }

    private string _Name;
    [Persistent("Name")]
    public string Name { get => _Name; set => SetPropertyValue(nameof(Name), ref _Name, value); }

    private int _HourSum;
    [Persistent("HourSum")]
    public int HourSum { get => _HourSum; set => SetPropertyValue(nameof(HourSum), ref _HourSum, value); }
}
```

> > **Important**: this only covers this very basic example, and highly depends on your configuration (e.g. SecuriySystem, Defered deletion etc.). You will need good testing on that to not get stale or out of sync.  

Reviewer: ```suggestion
> **Important**: this only covers this very basic example, and highly depends on your configuration (e.g. SecuritySystem, Deferred deletion etc.). You will need good testing on that to not get stale or out of sync.  
```

Database side foreignkeys can help to mark stale objects, as well as triggers to handle that on a database level.  

> Also this *simple* technique only works if you are the only on that writes into the database.  

Reviewer: ```suggestion
Also this *simple* technique only works if you are the only one that writes into the database.  
```

> I will cover mode advanced scenarios in the future, just let me know in the comments below.

Reviewer: ```suggestion
```

Repeat from above

> This will result in a slower write performance for the `Order` but a **huge** speed improvment on the `Query` table. I've set the `DataAccessMode` of the `Query` table to `DataView` and left the `DataAccessMode` for the `OrderTable` in `Client` mode:

Reviewer: ```suggestion
This will result in a slower write performance for the `Offer` but a **huge** speed improvement on the `Query` table. I've set the `DataAccessMode` of the `Query` table to `DataView` and left the `DataAccessMode` for the `OfferTable` in `Client` mode:
```

> `Order`:

Reviewer: ```suggestion
`Offer`:
```

```txt
19.09.20 13:39:21.249 Executing sql 'select N0."Oid",N0."Name",N0."OptimisticLockField",N0."GCRecord" from "dbo"."FasterOfferWithCQRS" N0 where N0."GCRecord" is null'
19.09.20 13:39:21.250 Result: rowcount = 300, total = 11996, N0.{Oid,Guid} = 4800, N0.{Name,String} = 4796, N0.{OptimisticLockField,Int32} = 1200, N0.{GCRecord,Int32} = 1200
19.09.20 13:39:21.264 Executing sql 'select top 1 count(*) from "dbo"."FasterOfferWithCQRS" N0 where N0."GCRecord" is null'
19.09.20 13:39:21.265 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
19.09.20 13:39:21.266 Executing sql 'select top 1 count(*) from "dbo"."FasterOfferItemWithCQRS" N0 where N0."GCRecord" is null'
19.09.20 13:39:21.309 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
```

This results in the following performance

1. Sort by `Name`: 0.014 seconds

> > *Hint*: we can still provide an `NonPersistent` `HourSum` field in the `Order` class, but make sure users can't display, filter, group by this field in any `ListView`. In `DetailViews` this should not be a huge performance penalty.

Reviewer: ```suggestion
> *Hint*: we can still provide an `NonPersistent` `HourSum` field in the `Offer` class, but make sure users can't display, filter, group by this field in any `ListView`. In `DetailViews` this should not be a huge performance penalty.
```

`Query`:

```txt
19.09.20 13:40:32.895 Executing sql 'select top 2147483647 N0."HourSum",N0."Name",N0."Oid" from "dbo"."FasterOfferWithCQRSQuery" N0 where N0."GCRecord" is null'
19.09.20 13:40:32.896 Result: rowcount = 300, total = 10796, N0.{HourSum,Int32} = 1200, N0.{Name,String} = 4796, N0.{Oid,Guid} = 4800
19.09.20 13:40:32.900 Executing sql 'select top 1 count(*) from "dbo"."FasterOfferWithCQRSQuery" N0 where N0."GCRecord" is null'
19.09.20 13:40:32.900 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
```

This results in the following performance

1. Sort by `Name`: 0.005 seconds
2. Sort by `HourSum`: 0.004 seconds

> **Important**: This will of course result in data duplication and use more disk space.  
Keep that in mind when designing your datacenter.  

> It also will slow down bulk inserts, cause a lot of more records need to be created.

Reviewer: ```suggestion
It also will slow down bulk inserts, because a lot of more records need to be created.
```

### Use the database at your advantage - Views

Until now every technique we used does not care about the database underneath. If you don't plan to support more than one or two databases (and be honest, when did you switch your database the last time?) use the **power** of your database.

> Databases are good at this. Thats their *job*. But we need to drop down to **SQL**. Don't be afraid, it has a lot of benifits on the long run.

Reviewer: ```suggestion
Databases are good at this. That is their *job*. But we need to drop down to **SQL**. Don't be afraid, it has a lot of benefits on the long run.
```

{
    //Register the updater in the module
    public override IEnumerable<ModuleUpdater> GetModuleUpdaters(IObjectSpace objectSpace, Version versionFromDB) => new ModuleUpdater[]
    {
        new CreateViewDatabaseUpdater(objectSpace, versionFromDB)
    };
}

```

`Order`:

```txt
19.09.20 14:45:18.378 Executing sql 'select N0."Oid",N0."Name",N0."OptimisticLockField",N0."GCRecord" from "dbo"."FasterOfferWithView" N0 where N0."GCRecord" is null'
19.09.20 14:45:18.380 Result: rowcount = 300, total = 12078, N0.{Oid,Guid} = 4800, N0.{Name,String} = 4878, N0.{OptimisticLockField,Int32} = 1200, N0.{GCRecord,Int32} = 1200
19.09.20 14:45:18.388 Executing sql 'select top 1 count(*) from "dbo"."FasterOfferWithView" N0 where N0."GCRecord" is null'
19.09.20 14:45:18.389 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
19.09.20 14:45:18.390 Executing sql 'select top 1 count(*) from "dbo"."FasterOfferItemWithView" N0 where N0."GCRecord" is null'
19.09.20 14:45:18.410 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
```

This results in the following performance

1. Sort by `Name`: 0.009 seconds

> > *Hint*: we can still provide an `NonPersistent` `HourSum` field in the `Order` class, but make sure users can't display, filter, group by this field in any `ListView`. In `DetailViews` this should not be a huge performance penalty.

Reviewer: ```suggestion
> *Hint*: we can still provide a `NonPersistent` `HourSum` field in the `Offer` class, but make sure users can't display, filter, group by this field in any `ListView`. In `DetailViews` this should not be a huge performance penalty.
```

`Query`:

```txt
19.09.20 14:46:07.290 Executing sql 'select top 2147483647 N0."HourSum",N0."Name",N0."Oid" from "dbo"."FasterOfferWithViewQuery" N0 where N0."GCRecord" is null'
19.09.20 14:46:07.356 Result: rowcount = 300, total = 10878, N0.{HourSum,Int32} = 1200, N0.{Name,String} = 4878, N0.{Oid,Guid} = 4800
19.09.20 14:46:07.360 Executing sql 'select top 1 count(*) from "dbo"."FasterOfferWithViewQuery" N0 where N0."GCRecord" is null'
19.09.20 14:46:07.362 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
```

This results in the following performance

1. Sort by `Name`: 0.079 seconds
2. Sort by `HourSum`: 0.070 seconds

> That's pretty impressiv! Finally let's have a look at the totally unsupported pitched version that saved me several times.

Reviewer: ```suggestion
That's very impressive! Finally let's have a look at the totally unsupported pitched version that saved me a lot of time.
```

Do you mean that you saved actual time: It saved me 10 hours or that it saved your life/ass many times?

Author: Both. 😂 But more the latter

### Totally unsupported - There will be dragons - N+N query version

> **WARNING**: THIS WILL AND CAN BREAK IN THE FUTURE!  
FROM THIS POINT ON YOU ARE ON YOUR OWN.

> What if we can not avoid N+1 queries, but at least come down to N+N queries? That means we let XAF use a normal `Client` mode ListView and do 1 additional query for all records in that perticular `ListView`?

Reviewer: ```suggestion
What if we can not avoid N+1 queries, but at least come down to N+N queries? That means we let XAF use a normal `Client` mode ListView and do 1 additional query for all records in that particular `ListView`?
```

> There is one not by XAF supported feature called `Session.Prefetch` but it has also some limitations. We need a way to do 1 query when the **first** N+1 query would occur, afterwards we cache it and just lookup data from this cache.  

Reviewer: ```suggestion
There is one in XAF unsupported features called `Session.Prefetch` but it also has some limitations. We need a way to do 1 query when the **first** N+1 query would occur, afterwards we cache it and just lookup data from this cache.  
```

Author: I am Not Sure, but i rephrased that whole Paragraph. Starting misstrusting the universe. Damn Mate icetea. Should stay with 🍺.

> We cant use static fields, cause we have no idea when to purge the cache. But there is one *undocumented* feature of XPO called `IWideDataStorage` we can leverage.

Reviewer: ```suggestion
We can't use static fields, cause we have no idea when to purge the cache. But there is one *undocumented* feature of XPO called `IWideDataStorage` we can leverage.
```

19.09.20 15:51:57.287 Result: rowcount = 300, total = 11950, N0.{Oid,Guid} = 4800, N0.{Name,String} = 4750, N0.{OptimisticLockField,Int32} = 1200, N0.{GCRecord,Int32} = 1200
19.09.20 15:51:57.289 Executing sql 'select N0."Oid",(select sum(N1."Hours") as Res0 from "dbo"."FasterOfferItem" N1 where ((N0."Oid" = N1."FasterOffer") and N1."GCRecord" is null)) from "dbo"."FasterOffer" N0 where N0."GCRecord" is null'
19.09.20 15:51:57.359 Result: rowcount = 300, total = 6000, N0.{Oid,Guid} = 4800, SubQuery(Sum,N1.{Hours,Int32},Select N1.{Hours,Int32}
  from "FasterOfferItem" N1
 where N0.{Oid,Guid} = N1.{FasterOffer,Guid} And N1.{GCRecord,Int32} Is Null) = 1200
19.09.20 15:51:57.363 Executing sql 'select top 1 count(*) from "dbo"."FasterOffer" N0 where N0."GCRecord" is null'
19.09.20 15:51:57.363 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
19.09.20 15:51:57.364 Executing sql 'select top 1 count(*) from "dbo"."FasterOfferItem" N0 where N0."GCRecord" is null'
19.09.20 15:51:57.387 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4

```

This results in the following performance

1. Sort by `Name`: 0.082 seconds
2. Sort by `HourSum`: 0.076 seconds

> > **Note**: Cause XAF recreates a `Session` object every time it's refreshed, will fetch the second query only once in the livetime of the session.  

Reviewer: ```suggestion
> **Note**: Because XAF recreates a `Session` object every time it's refreshed, will fetch the second query only once in the lifetime of the session.  
```

> You can control the cache what ever way you like.  

Reviewer: ```suggestion
You can control the cache whichever way you like. 
```

> Beware that you will pay the cost of calculating the aggregates of ALL objects in the `DetailView` context as well (cause there is no natural way to figure out if you are currently displayed in a `DetailView`).  

Reviewer: ```suggestion
Beware that you will pay the cost of calculating the aggregates of ALL objects in the `DetailView` context as well (because there is no natural way to figure out if you are currently displayed in a `DetailView`).  
```

> There are several strategies you can further improve this technique, but for now I think this is pretty impressive.

Reviewer: ```suggestion
There are several strategies you can further improve using this technique, but for now I think this is pretty impressive.
```

> This is really awesome performance for very little effort. Of course you can combine those techniques. Use `XPQuery` combined with `XPView` and so on.

Reviewer: ```suggestion
This is really awesome performance gain for very little effort. Of course you can combine those techniques. Use `XPQuery` combined with `XPView` and so on.
```

The main goal of this post is how to identify performance bottlenecks in your application and how to overcome them when dealing with aggregates especially in `ListViews`.

### Summary - Lessions learned

1. Performance is hard
1. Measure, measure, measure
1. Pick the right tool (and always keep memory and database load in sight)
1. Never combine `PersistentAlias` with client mode aggregates if possible
1. Use your database as a tool
1. If everything performance wise breaks down: use and measure the last 3 options

> 1. Use N+N query wisely. It doesn't have the same befinits like all the other `DataAccessModes` but it behaves linear, and can help calculate complicated business rules only once and avoid the N+1 problem.

Reviewer: ```suggestion
1. Use N+N query wisely. It doesn't have the same benefits like all the other `DataAccessModes` but it behaves linear, and can help calculate complicated business rules only once and avoid the N+1 problem.
```

1. Everything is a tradeoff (implementation time, memory, cpu time, stale data)

> 1. Use `DataView` or `InstantFeedbackView` in combination with `PersistentAlias` where ever you will need to do aggregation with larger amounts of data

Reviewer: ```suggestion
1. Use `DataView` or `InstantFeedbackView` in combination with `PersistentAlias` wherever you need to do aggregation with larger amounts of data
```

> 1. `ServerMode`, `ServerView` and `InstantFeedback` will drive your `DBA` crazy, if used uncorrectly

Reviewer: ```suggestion
1. `ServerMode`, `ServerView` and `InstantFeedback` will drive your `DBA` crazy, if used incorrectly
```

1. Aim for UX first and stay with `Client` mode as much as you can

> 1. You really need good reasons for `CQRS`. It adds **loads** of performance, but increases complexity and maintainance a **lot**

Reviewer: ```suggestion
1. You really need good reasons for `CQRS`. It adds **loads** of performance, but increases complexity and maintenance a **lot**
```

> 1. Database `VIEWS` are cheaper from maintainance perspective than `CQRS`

Reviewer: ```suggestion
1. Database `VIEWS` are cheaper from maintenance perspective than `CQRS`
```

I didn't even dig into execution plans or something special database wise. That is totally out of scope of this post.  
One thing I always recomend is: stick with the `Client` `DataAccessMode` as long as you can, esp. for smaller record sets. It will perform really well, if you keep an eye on *chatty* requests (N+1).

> We had no `Indexes` what so ever (except the default ones XPO creates for us). Databases are pretty damn fast.

Reviewer: ```suggestion
We had no `Indexes` whatsoever (except the default ones XPO creates for us). Databases are pretty damn fast.
```

19.09.20 15:51:57.287 Result: rowcount = 300, total = 11950, N0.{Oid,Guid} = 4800, N0.{Name,String} = 4750, N0.{OptimisticLockField,Int32} = 1200, N0.{GCRecord,Int32} = 1200
19.09.20 15:51:57.289 Executing sql 'select N0."Oid",(select sum(N1."Hours") as Res0 from "dbo"."FasterOfferItem" N1 where ((N0."Oid" = N1."FasterOffer") and N1."GCRecord" is null)) from "dbo"."FasterOffer" N0 where N0."GCRecord" is null'
19.09.20 15:51:57.359 Result: rowcount = 300, total = 6000, N0.{Oid,Guid} = 4800, SubQuery(Sum,N1.{Hours,Int32},Select N1.{Hours,Int32}
  from "FasterOfferItem" N1
 where N0.{Oid,Guid} = N1.{FasterOffer,Guid} And N1.{GCRecord,Int32} Is Null) = 1200
19.09.20 15:51:57.363 Executing sql 'select top 1 count(*) from "dbo"."FasterOffer" N0 where N0."GCRecord" is null'
19.09.20 15:51:57.363 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
19.09.20 15:51:57.364 Executing sql 'select top 1 count(*) from "dbo"."FasterOfferItem" N0 where N0."GCRecord" is null'
19.09.20 15:51:57.387 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4

```

This results in the following performance

1. Sort by `Name`: 0.920 seconds
2. Sort by `HourSum`: 0.876 seconds

### Recap

> There is no one size fit's it all. Performance will always be hard. But I hope you learned some techniques to measure an improve your applications performance.

Reviewer: ```suggestion
There is no one size fits it all. Performance will always be hard. But I hope you learned some techniques to measure and improve your applications performance.
```

If you find interesting what I'm doing, consider becoming a [patreon](//www.patreon.com/biohaz999) or [contact me](//www.delegate.at/) for training, development or consultancy.

> **New Kid on the block**: You now can support me on several channels for all kind of projects. Head over to my [new baby called Tasty](https://tasty.xenial.io/support/). [Tasty](https://tasty.xenial.io) is a delicious dotnet testing platform you can use with and in any application. I would be more than happy if you support me and the project.  
There will be a new page for [Xenial](https://www.xenial.io/) soon. The project with the pure goal to make you even be **more** productive with XAF, XPO and all business related development.

Stay awesome!  
Manuel

> > PLEASE NEVER DO SOMETHING LIKE THIS IN PRODUCTION. YOU HAVE BEEN WARNED. (Or I insist you NEED to book some consulting hour's from me)

Reviewer: ```suggestion
```

Repeat from above
